pkg/retrieval/retrieval.go

```diff
@@ -126,156 +126,145 @@ func (s *Service) RetrieveChunk(ctx context.Context, addr, sourcePeerAddr swarm.
-			if peersResults >= maxPeers {
-				loggerV1.Debug("failed to get chunk", "chunk_address", addr)
-				return nil, storage.ErrNotFound
-			}
 
-			// if we have not counted enough successful attempts but out of selection amount, reset
-			if peerAttempt >= maxSelects {
-
-				// forwarder gives up after first round of request attempts,
-				// rather than holding up the forwarding chain by waiting to be able to reserve downstream peer
-				if !origin {
-					return nil, storage.ErrNotFound
-				}
-
-				requestAttempt++
-				timeNow := time.Now().Unix()
-				if timeNow > lastTime {
-					lastTime = timeNow
-					peerAttempt = 0
+					// there are overdrafted peers, so we want to retry with them again
+					loggerV1.Debug("peers are overdrafted, retrying", "chunk_address", addr)
 					sp.ResetOverdraft()
-				} else {
 					select {
-					case <-time.After(600 * time.Millisecond):
+					case <-time.After(resetOverdraftDur): // wait at least 600 milliseconds
+						retry()
+						continue
 					case <-ctx.Done():
-						loggerV1.Debug("failed to get chunk", "chunk_address", addr, "error", ctx.Err())
-						return nil, fmt.Errorf("retrieval: %w", ctx.Err())
+						return nil, ctx.Err()
 					}
 				}
-			}
 
+				if res.retrieveAttempted {
+					retrievedErrorsLeft--
+				}
+
+				if res.err != nil {
+					loggerV1.Debug("failed to get chunk", "chunk_address", addr, "peer_address", res.peer, "error", res.err)
+					retry()
+				} else {
+					loggerV1.Debug("retrieved chunk", "chunk_address", addr, "peer_address", res.peer)
+					return res.chunk, nil
+				}
+			}
 		}
 
-		// if we have not managed to get results after 5 (maxRequestRounds) rounds of peer selections, give up
+		loggerV1.Debug("no peers left to retry", "chunk_address", addr)
 		return nil, storage.ErrNotFound
-
 	})
 	if err != nil {
+		s.logger.Debug("retrieval failed", "chunk_address", addr, "error", err)
 		return nil, err
 	}
 
 	return v.(swarm.Chunk), nil
 }
 
-func (s *Service) retrieveChunk(ctx context.Context, addr swarm.Address, sp *skippeers.List, isOrigin bool) (chunk swarm.Chunk, peer swarm.Address, requested bool, err error) {
-	loggerV1 := s.logger.V(1).Build()
+func (s *Service) retrieveChunk(ctx context.Context, done chan struct{}, result chan retrievalResult, addr swarm.Address, sp *skippeers.List, isOrigin bool) {
 
 	startTimer := time.Now()
 	// allow upstream requests if this node is the source of the request
 	// i.e. the request was not forwarded, to improve retrieval
 	// if this node is the closest to he chunk but still does not contain it
 	allowUpstream := isOrigin
 
+	var (
+		err               error
+		peer              swarm.Address
+		retrieveAttempted bool
+		chunk             swarm.Chunk
+	)
+
+	defer func() {
+		select {
+		case result <- retrievalResult{err: err, chunk: chunk, retrieveAttempted: retrieveAttempted, peer: peer}:
+		case <-done:
+			return
+		}
+	}()
+
+	defer func() {
```

Comment:
No need for a second `defer` call, the error check can be put at the beginning of the `defer` above.